soylent.com: dropdown menu dismisses when trying to tap on links

whsieh opened the pull request:
#### ea5905fd9777612df86ae3b8471c24c60e9935b7
<pre>
soylent.com: dropdown menu dismisses when trying to tap on links
<a href="https://bugs.webkit.org/show_bug.cgi?id=282476">https://bugs.webkit.org/show_bug.cgi?id=282476</a>
<a href="https://rdar.apple.com/113314067">rdar://113314067</a>

Reviewed by Aditya Keerthi.

On soylent.com, the site navigation dropdown contains a `pointerout` event listener which dismisses
the entire dropdown when dispatched. This, when combined with synthetic mouse events and content
observation, makes it impossible to actually navigate to any links in the header, since we get the
following sequence of events:

a.  User begins a touch over a link inside the dropdown menu, dispatching `pointerdown`
b.  User ends the touch, dispatching `pointerup`
c.  Because the touch has ended, we immediately dispatch `pointerout`, and the page closes the menu
d.  In parallel, a synthetic click is recognized in the UI process, dispatching a synthetic click to
    the page
e.  This synthetic click hits the `body` element (because the dropdown menu has already closed), and
    we don&apos;t click the link

To fix this, we add a quirk to ensure that step (c) now happens after the synthetic click is
dispatched to the page on `soylent.com`; additionally, this quirk *prevents* the `pointerout` event
from being dispatched after ending the touch, in the case where the synthetic click resulted in a
nontrivial content observation (such that the synthetic click state remains in hover state). This
latter adjustment is required in order to prevent the dropdown menu from immediately closing after
opening the menu, as a result of dispatching `pointerout`.

At a high level, this bug boils down to the fact that when using a touch-capable device, `pointer*`
events track the user&apos;s touch state and follow the life cycle of touch events, rather than mapping
to the compatibility mouse events (synthetic clicks) that are dispatched in response to tapping
elements on the page. soylent.com presents an interesting situation where the webpage relies on
*both* compatibility mouse events (so that the menu opens when tapping on the dropdown button), as
well as pointer events (for knowing when to dismiss the menu). The behavior introduced in this quirk
essentially bridges this gap by allowing `pointerout` to honor the synthetic mouse hover and clicks
dispatched when tapping the screen, but is likely too risky to deploy in the general case.

* LayoutTests/pointerevents/ios/dispatch-pointerout-after-synthetic-click-quirk-expected.txt: Added.
* LayoutTests/pointerevents/ios/dispatch-pointerout-after-synthetic-click-quirk.html: Added.

Add a layout test to exercise this quirk.

* Source/WebCore/Headers.cmake:

Add `SyntheticClickResult.h`.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:

Update Safer C++ expectations, now that `m_page` is a weak pointer.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::callAfterPendingSyntheticClick):

Add a chrome client hook to invoke a callback after waiting for a pending synthetic click completes.
This may be invoked immediately or asynchronously, with the following results:

• `Failed`        The gesture was not recognized as a synthetic click
• `Click`         The synthetic click was dispatched
• `Hover`         The synthetic click stopped at mouse hover state
• `PageInvalid`   The page was invalidated (either because the tab was closed, or we navigated away)

In the case of `Hover` and `PageInvalid`, we avoid dispatching the `pointer(out|leave)` events.

* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchOverOrOutEvent):
(WebCore::PointerCaptureController::dispatchEnterOrLeaveEvent):

Pull these two helpers out as private methods on `PointerCaptureController`, so we can call into
them asynchronously below.

(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):

Use `callAfterPendingSyntheticClick` to defer the `pointerout` and `pointerleave` events until after
the synthetic click has completed (and even withhold them altogether, in the case of synthetic
hover).

(WebCore::PointerCaptureController::cancelPointer):
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::isYahooMail const):
(WebCore::Quirks::isDomain const):
(WebCore::Quirks::domainStartsWith const):
(WebCore::Quirks::needsFormControlToBeMouseFocusable const):
(WebCore::Quirks::isTouchBarUpdateSuppressedForHiddenContentEditable const):
(WebCore::Quirks::isNeverRichlyEditableForTouchBar const):
(WebCore::Quirks::shouldSuppressAutocorrectionAndAutocapitalizationInHiddenEditableAreas const):
(WebCore::Quirks::shouldDisableWritingSuggestionsByDefault const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldPreventDispatchOfTouchEvent const):
(WebCore::Quirks::shouldAvoidResizingWhenInputViewBoundsChange const):
(WebCore::Quirks::needsDeferKeyDownAndKeyPressTimersUntilNextEditingCommand const):
(WebCore::Quirks::shouldBypassBackForwardCache const):
(WebCore::Quirks::shouldBypassAsyncScriptDeferring const):
(WebCore::Quirks::shouldAvoidPastingImagesAsWebContent const):
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
(WebCore::Quirks::requiresUserGestureToPauseInPictureInPicture const):
(WebCore::Quirks::shouldDisableEndFullscreenEventWhenEnteringPictureInPictureFromFullscreenQuirk const):
(WebCore::Quirks::allowLayeredFullscreenVideos const):
(WebCore::Quirks::shouldPreventOrientationMediaQueryFromEvaluatingToLandscape const):
(WebCore::Quirks::shouldFlipScreenDimensions const):
(WebCore::Quirks::shouldIgnoreTextAutoSizing const):
(WebCore::Quirks::scriptToEvaluateBeforeRunningScriptFromURL):
(WebCore::Quirks::shouldHideCoarsePointerCharacteristics const):
(WebCore::Quirks::implicitMuteWhenVolumeSetToZero const):
(WebCore::Quirks::shouldDispatchPointerOutAfterHandlingSyntheticClick const):
(WebCore::Quirks::needsMozillaFileTypeForDataTransfer const):
(WebCore::Quirks::setTopDocumentURLForTesting):
(WebCore::Quirks::topDocumentURL const):

Add the quirk, and also add plumbing for a testing-only hook to allow tests that have access to
`window.internals` to override the top document URL used for deciding quirks. The new layout test
uses this mechanism.

(WebCore::isYahooMail): Deleted.

Turn this into a method instead, so that we can use `topDocumentURL` which accounts for the new
override URL for testing.

* Source/WebCore/page/Quirks.h:
* Source/WebCore/platform/SyntheticClickResult.h: Added.

Add a new enum type; see above for more details.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setTopDocumentURLForQuirks):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add a new testing hook to simulate top URLs for quirks; see above for more details.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

Add plumbing for an IPC message that returns whether or not a potential tap is in progress on iOS.
See below for more details.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::isPotentialTapInProgress const):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView isPotentialTapInProgress]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::isPotentialTapInProgress):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::callAfterPendingSyntheticClick):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::didCommitLoad):

Invoke the pending synthetic click callback with `PageInvalid`, if needed.

(WebKit::WebPage::callAfterPendingSyntheticClick):

Add an empty stub for non-iOS platforms.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):
(WebKit::WebPage::didHandleTapAsHover):
(WebKit::WebPage::didFinishContentChangeObserving):
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::commitPotentialTapFailed):

Add calls to invoke the pending synthetic click callback throughout the various checkpoints during
synthetic mouse event dispatch: `Failed` (for when the potential tap node is rejected), `Hover` (for
when content change observation fires), and `Click` (for when the click is dispatched).

(WebKit::WebPage::invokePendingSyntheticClickCallback):
(WebKit::WebPage::callAfterPendingSyntheticClick):

Implement a new helper method on `WebPage` that invokes the callback after the current synthetic
click event has been dispatched. This avoids racing against gesture state in the UI process, by
round-tripping to `WKContentView` to determine whether or not the single tap gesture recognizer has
fired before stashing the callback in `m_pendingSyntheticClickCallback` on the `WebPage`.

Canonical link: <a href="https://commits.webkit.org/286270@main">https://commits.webkit.org/286270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1030e832154509eebf2e53de76a73300d1339aa0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75377 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/39 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79851 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26641 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77493 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/39 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2608 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59158 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17373 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78444 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/39 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64759 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39520 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/39 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22259 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24968 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/39 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22592 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81333 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2716 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2867 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64746 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/66684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16605 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10636 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8797 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2673 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5494 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2698 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3633 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2705 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/pointerevents/ios/dispatch-pointerout-after-synthetic-click-quirk.html:
```diff
@@ -0,0 +1,52 @@
+window.internals?.setTopDocumentURLForQuirks("https://soylent.com");
+
+addEventListener("load", async () => {
+    description("Verifies that tapping on a link inside of a container that hides itself when receiving . This test requires WebKitTestRunner");
```
when receiving pointerleave?

aprotyas replied:
i.e. I think there's a missing word
> "Verifies that tapping on a link inside of a container that hides itself when receiving . This

whsieh replied:
Heh, yeah I accidentally a word 😓.

Added `pointerleave` here, thanks!

whsieh committed:
soylent.com: dropdown menu dismisses when trying to tap on links
https://bugs.webkit.org/show_bug.cgi?id=282476
rdar://113314067

Reviewed by Aditya Keerthi.

On soylent.com, the site navigation dropdown contains a `pointerout` event listener which dismisses
the entire dropdown when dispatched. This, when combined with synthetic mouse events and content
observation, makes it impossible to actually navigate to any links in the header, since we get the
following sequence of events:

a.  User begins a touch over a link inside the dropdown menu, dispatching `pointerdown`
b.  User ends the touch, dispatching `pointerup`
c.  Because the touch has ended, we immediately dispatch `pointerout`, and the page closes the menu
d.  In parallel, a synthetic click is recognized in the UI process, dispatching a synthetic click to
    the page
e.  This synthetic click hits the `body` element (because the dropdown menu has already closed), and
    we don't click the link

To fix this, we add a quirk to ensure that step (c) now happens after the synthetic click is
dispatched to the page on `soylent.com`; additionally, this quirk *prevents* the `pointerout` event
from being dispatched after ending the touch, in the case where the synthetic click resulted in a
nontrivial content observation (such that the synthetic click state remains in hover state). This
latter adjustment is required in order to prevent the dropdown menu from immediately closing after
opening the menu, as a result of dispatching `pointerout`.

At a high level, this bug boils down to the fact that when using a touch-capable device, `pointer*`
events track the user's touch state and follow the life cycle of touch events, rather than mapping
to the compatibility mouse events (synthetic clicks) that are dispatched in response to tapping
elements on the page. soylent.com presents an interesting situation where the webpage relies on
*both* compatibility mouse events (so that the menu opens when tapping on the dropdown button), as
well as pointer events (for knowing when to dismiss the menu). The behavior introduced in this quirk
essentially bridges this gap by allowing `pointerout` to honor the synthetic mouse hover and clicks
dispatched when tapping the screen, but is likely too risky to deploy in the general case.

* LayoutTests/pointerevents/ios/dispatch-pointerout-after-synthetic-click-quirk-expected.txt: Added.
* LayoutTests/pointerevents/ios/dispatch-pointerout-after-synthetic-click-quirk.html: Added.

Add a layout test to exercise this quirk.

* Source/WebCore/Headers.cmake:

Add `SyntheticClickResult.h`.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:

Update Safer C++ expectations, now that `m_page` is a weak pointer.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::callAfterPendingSyntheticClick):

Add a chrome client hook to invoke a callback after waiting for a pending synthetic click completes.
This may be invoked immediately or asynchronously, with the following results:

• `Failed`        The gesture was not recognized as a synthetic click
• `Click`         The synthetic click was dispatched
• `Hover`         The synthetic click stopped at mouse hover state
• `PageInvalid`   The page was invalidated (either because the tab was closed, or we navigated away)

In the case of `Hover` and `PageInvalid`, we avoid dispatching the `pointer(out|leave)` events.

* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchOverOrOutEvent):
(WebCore::PointerCaptureController::dispatchEnterOrLeaveEvent):

Pull these two helpers out as private methods on `PointerCaptureController`, so we can call into
them asynchronously below.

(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):

Use `callAfterPendingSyntheticClick` to defer the `pointerout` and `pointerleave` events until after
the synthetic click has completed (and even withhold them altogether, in the case of synthetic
hover).

(WebCore::PointerCaptureController::cancelPointer):
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::isYahooMail const):
(WebCore::Quirks::isDomain const):
(WebCore::Quirks::domainStartsWith const):
(WebCore::Quirks::needsFormControlToBeMouseFocusable const):
(WebCore::Quirks::isTouchBarUpdateSuppressedForHiddenContentEditable const):
(WebCore::Quirks::isNeverRichlyEditableForTouchBar const):
(WebCore::Quirks::shouldSuppressAutocorrectionAndAutocapitalizationInHiddenEditableAreas const):
(WebCore::Quirks::shouldDisableWritingSuggestionsByDefault const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldPreventDispatchOfTouchEvent const):
(WebCore::Quirks::shouldAvoidResizingWhenInputViewBoundsChange const):
(WebCore::Quirks::needsDeferKeyDownAndKeyPressTimersUntilNextEditingCommand const):
(WebCore::Quirks::shouldBypassBackForwardCache const):
(WebCore::Quirks::shouldBypassAsyncScriptDeferring const):
(WebCore::Quirks::shouldAvoidPastingImagesAsWebContent const):
(WebCore::Quirks::requestStorageAccessAndHandleClick const):
(WebCore::Quirks::requiresUserGestureToPauseInPictureInPicture const):
(WebCore::Quirks::shouldDisableEndFullscreenEventWhenEnteringPictureInPictureFromFullscreenQuirk const):
(WebCore::Quirks::allowLayeredFullscreenVideos const):
(WebCore::Quirks::shouldPreventOrientationMediaQueryFromEvaluatingToLandscape const):
(WebCore::Quirks::shouldFlipScreenDimensions const):
(WebCore::Quirks::shouldIgnoreTextAutoSizing const):
(WebCore::Quirks::scriptToEvaluateBeforeRunningScriptFromURL):
(WebCore::Quirks::shouldHideCoarsePointerCharacteristics const):
(WebCore::Quirks::implicitMuteWhenVolumeSetToZero const):
(WebCore::Quirks::shouldDispatchPointerOutAfterHandlingSyntheticClick const):
(WebCore::Quirks::needsMozillaFileTypeForDataTransfer const):
(WebCore::Quirks::setTopDocumentURLForTesting):
(WebCore::Quirks::topDocumentURL const):

Add the quirk, and also add plumbing for a testing-only hook to allow tests that have access to
`window.internals` to override the top document URL used for deciding quirks. The new layout test
uses this mechanism.

(WebCore::isYahooMail): Deleted.

Turn this into a method instead, so that we can use `topDocumentURL` which accounts for the new
override URL for testing.

* Source/WebCore/page/Quirks.h:
* Source/WebCore/platform/SyntheticClickResult.h: Added.

Add a new enum type; see above for more details.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setTopDocumentURLForQuirks):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add a new testing hook to simulate top URLs for quirks; see above for more details.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

Add plumbing for an IPC message that returns whether or not a potential tap is in progress on iOS.
See below for more details.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::isPotentialTapInProgress const):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView isPotentialTapInProgress]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::isPotentialTapInProgress):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::callAfterPendingSyntheticClick):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::didCommitLoad):

Invoke the pending synthetic click callback with `PageInvalid`, if needed.

(WebKit::WebPage::callAfterPendingSyntheticClick):

Add an empty stub for non-iOS platforms.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):
(WebKit::WebPage::didHandleTapAsHover):
(WebKit::WebPage::didFinishContentChangeObserving):
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::commitPotentialTapFailed):

Add calls to invoke the pending synthetic click callback throughout the various checkpoints during
synthetic mouse event dispatch: `Failed` (for when the potential tap node is rejected), `Hover` (for
when content change observation fires), and `Click` (for when the click is dispatched).

(WebKit::WebPage::invokePendingSyntheticClickCallback):
(WebKit::WebPage::callAfterPendingSyntheticClick):

Implement a new helper method on `WebPage` that invokes the callback after the current synthetic
click event has been dispatched. This avoids racing against gesture state in the UI process, by
round-tripping to `WKContentView` to determine whether or not the single tap gesture recognizer has
fired before stashing the callback in `m_pendingSyntheticClickCallback` on the `WebPage`.

Canonical link: https://commits.webkit.org/286270@main
